migen/build/platforms/tinyfpga_a.py

```diff
@@ -0,0 +1,99 @@
+import bisect
+
+from migen import *
+from migen.build.generic_platform import *
+from migen.build.lattice import LatticePlatform
+
+_io = [
+]
+
+_connectors = [
+    # Pins 1-22
+    ("GPIO", "13 14 16 17 20 21 23 25 26 27 28 4 5 8 9 10 11 12"),
+]
+
+
+# Default peripherals.
+# Only on AX.
+led = [
+    ("user_led", 0, Pins("GPIO:0"))
+]
+
+serial = [
+    ("serial", 0,
+        Subsignal("tx", Pins("GPIO:2")),
+        Subsignal("rx", Pins("GPIO:3")),
+        IOStandard("LVCMOS33")
+     )
+]
+
+
+class MachClock(Module):
+    # Diamond will complain that the frequency of the oscillator doesn't match
+    # if it doesn't match one of the below frequencies. Since we tend to deal
+    # with periods in nanoseconds, choose the closest frequency (in MHz)
+    # without going over. These numbers were extracted from
+    # "MachXO2 sysCLOCK PLL Design and Usage Guide"
+    supported_freqs = [
+        2.08, 2.15, 2.22, 2.29, 2.38, 2.46, 2.56, 2.66, 2.77, 2.89,
+        3.02, 3.17, 3.33, 3.50, 3.69, 3.91, 4.16, 4.29, 4.43, 4.59,
+        4.75, 4.93, 5.12, 5.32, 5.54, 5.78, 6.05, 6.33, 6.65, 7.00,
+        7.39, 7.82, 8.31, 8.58, 8.87, 9.17, 9.50, 9.85, 10.23, 10.64,
+        11.08, 11.57, 12.09, 12.67, 13.30, 14.00, 14.78, 15.65, 15.65, 16.63,
+        17.73, 19.00, 20.46, 22.17, 24.18, 26.60, 29.56, 33.25, 38.00, 44.33,
+        53.20, 66.50, 88.67, 133.00
+    ]
+
+    def __init__(self, period, out):
+        self.specials += Instance("OSCH",
+                                  p_NOM_FREQ=str(self.nearest_freq(period)),
+                                  i_STDBY=C(0),
+                                  o_OSC=out
+                                  )
+
+    def nearest_freq(self, period):
+        i = bisect.bisect_right(self.supported_freqs, 1000.0/period)
+        if i:
+            return self.supported_freqs[i-1]
+        raise ValueError("Clock period out of range for internal oscillator.")
+
```

Comment:
It's probably better computing the maximum error for each of the elements of the table and just select the frequency that  has the minimum error under +-5%.

Something like this:
````python
def nearest_supported_freq(freq):
    # compute error for each suported freq
    errors = [abs(f*1e6 - freq)/(f*1e6) for f in supported_freqs]
    print(errors)
    # select frequency in the +-5% range with the minimal error
    error = 0.05
    freq = None
    for f, e in zip(supported_freqs, errors):
        if e < error:
            error = e
            freq = f
    return freq
````

With the 66.5MHz setting, the oscillator can vary from 63.17Mhz to 69.82MHz.
With the 88.67Mhz setting, the oscillator can vary from 84.24Mhz to 93.10MHz.

With your algo, if user sets the frequency to 88.66MHz, we will select the 66.5Mhz setting, which is ok for the clock constraint but clearly not the best for the frequency accuracy.

When using the internal oscillator, the frequency can vary up to +-5%  and we don't need to be strict on the computations. We should probably just select the best setting in the +-5% range and use the +5% value for the clock constraint.

Reply:
Precomputing all errors is unneeded, hard to read and inefficient. Bisection is perfect to find the two best frequencies. Then decide which is the better based on whatever metric you determine to be best. It is definitely not obvious that smallest relative error is best. I can argue just as well for smallest absolute error or max frequency that's less or equal to the desired one.

Reply:
Exactly. Since you don't really worry so much about the frequency the plausible assumption is that you still care about timing closure. And then you want the frequency that's closest to your desired one. Ergo `bisect_right` is what you want.
The columns don't matter at all. It's just a monotonic list of frequencies. Why would you look at certain columns?

Reply:
Exactly. Since you don't really worry so much about the frequency the plausible assumption is that you still care about timing closure and you don't want this choice to be the reason for failure. And then you want the frequency that's closest to your desired one. Ergo bisect_right is what you want.
The columns don't matter at all. It's just a monotonic list of frequencies. Why would you look at certain columns?

Reply:
If my last argument does not make sense to you, then ok. (i don't really care about bisect or not, that just that i found the behaviour strange from a user perspective).
Now if you argue about timing closure, then the implemented code needs to be fixed:
Default frequency is 16MHz, code will select 15.65Mhz setting and constraint the design to 16Mhz  where datasheet say it can be up to 15.65*1.05=16.43MHz.

Reply:
Correct. Constraints should account for oscillator tolerance (if the timing engine doesn't already do that).
My main complaint was about the approach of computing unrelated errors and walking through the list with a lot of code while already knowing exactly which two values are closest.
The UART example is problematic in general independent from the algorithm: You'll have to work with the actual frequency anyway when determining the UART divider as the frequencies are spaced by much more than the UART tolerance would allow and 5% by itself is marginal.

Reply:
>Since you don't really worry so much about the frequency the plausible assumption is that you still care about timing closure and you don't want this choice to be the reason for failure.

(I don't see why one would choose a freq with the smallest absolute error. I don't think highest frequency less than `default_clock_period` is ideal, but choosing the frequency with the smallest relative error- when oscillator tolerance is taken into account- could mean that your chosen desired frequency _is_ the reason for timing closure failure.

> You'll have to work with the actual frequency anyway when determining the UART divider
I should probably include a helper method or member to return the actual chosen clock frequency (ideal) for cases like this.

Reply:
OK, then do we agree on that:
- use bisect to find which two values are closest.
- check if the requested frequency is in the +-5% range of the two values.
- if it's in none of them, then generate an error.
- if it's in one of them, then select this one.
- if it's in two of them, select the lower one.
- use selected freq + 5% for the clock constraints. (I don't know if the tool is already adding this 5% (vs something like 100x less for a typical oscillator)).
This way we know that returned freq will be in the ~+-5% range, and for the example of UART, it should work even if divider are computed with ideal requested frequency.

Reply:
> * use bisect to find which two values are closest.
> * check if the requested frequency is in the +-5% range of the two values.
> * if it's in none of them, then generate an error.

Sounds good to me.

> * if it's in one of them, then select this one.
> * if it's in two of them, select the lower one.

See below, but makes sense to me.

>use selected freq + 5% for the clock constraints.

Good idea. I bisected right- i.e. used less-than-or-equal under- the assumption that "the user puts in the highest frequency (clock period) they need the design to operate at, and are okay getting a _lower_ frequency", ignoring 5% oscillator error (though, like @jordens said, this isn't necessarily the only appropriate metric).

If we take 5% (well, any error really) error into account, the above assumption no longer holds, since the actual constraint emitted will always be for a higher frequency than the user specified. But that's okay; if your design doesn't pass timing at up to %5 higher frequency (up to 10% error if the higher of the two `bisect_right` values was chosen- _I'll put a comment for this_.), then your design at your desired frequency in the worst case wasn't going to work anyway. This board is meant to be low-cost, and using an internal oscillator is a tradeoff.

> (I don't know if the tool is already adding this 5% (vs something like 100x less for a typical oscillator)).

I can't find anything suggesting Diamond takes this into account in Lattice's docs. Assume for the tool doesn't use the 5%, if I find out later that Diamond _does_ account for this, I can send another PR.